_posts/2016-04-26-distributed-game-of-life-in-elixir.md

```diff
@@ -0,0 +1,1065 @@
+
+Our main supervisor is `GameOfLife.Supervisor` which I mentioned at the begining of the article. Here you can see how we defined its childrens like `Task.Supervisor`, workers for `BoardServer` and `GamePrinter`.
+
+{% highlight elixir %}
+# lib/game_of_life.ex
+defmodule GameOfLife do
+  use Application
+
+  # See http://elixir-lang.org/docs/stable/elixir/Application.html
+  # for more information on OTP Applications
+  def start(_type, _args) do
+    import Supervisor.Spec, warn: false
+
+    init_alive_cells = []
+
+    children = [
+      # Define workers and child supervisors to be supervised
+      # worker(GameOfLife.Worker, [arg1, arg2, arg3]),
+      supervisor(Task.Supervisor, [[name: GameOfLife.TaskSupervisor]]),
+      worker(GameOfLife.BoardServer, [init_alive_cells]),
+      worker(GameOfLife.GamePrinter, []),
+    ]
+
+    # See http://elixir-lang.org/docs/stable/elixir/Supervisor.html
+    # for other strategies and supported options
+    opts = [strategy: :one_for_one, name: GameOfLife.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+{% endhighlight %}
+
+Let me describe you what each component on the image is responsible for.
+
+* `Task.Supervisor` is Elixir module defining a new supervisor which can be used to dynamically supervise tasks. We are going to use it to spin off tasks like determining if the particular cell should live or die. Those tasks can be run across nodes connected into the cluster. In above code, we gave name `GameOfLife.TaskSupervisor` for our supervisor. We will use this name to tell `Task.Supervisor.async` function which Task Supervisor should handle our task. You can read more about [Task.Supervisor here](http://elixir-lang.org/docs/stable/elixir/Task.Supervisor.html).
+
+* `GameOfLife.BoardServer` is our module implemented as [GenServer behaviour](http://elixir-lang.org/docs/stable/elixir/GenServer.html). It's responsible for holding the state of the game. By that I mean it keeps the list of alive cells on the board along with generation counter and TRef. TRef is a timer reference returned by [Erlang timer module](http://erlang.org/doc/man/timer.html) and [apply_interval](http://erlang.org/doc/man/timer.html#apply_interval-4) function. We want to start the game and generate a new list of alive cells for next generation with specified time interval. With each new generation, we will update generation counter. The other interesting thing is that `GameOfLife.BoardServer` is running only on a single node. Once another node is connected to the cluster where is already running `GameOfLife.BoardServer` then `GameOfLife.BoardServer` won't be started just like that on the newly connected node. Instead on the new node `GameOfLife.BoardServer` will keep the only reference to the PID of the process existing on the first node. We want to have the single source of truth about the state of our game in one master `GameOfLife.BoardServer` process existing on the first node started in the cluster.
```

Comment:
With each new generation we will update generation counter, (without first coma)